### PR TITLE
feat: refactor the tag route feature with go

### LINKED ISSUE #1029: Refactor the tag route feature with go
- [x] I have searched the [issues](https://github.com/apache/dubbo-admin/issues) of this repository and believe that this is not a duplicate.
- [x] I have checked the [REAMDE](https://github.com/apache/dubbo-admin/blob/develop/README.md) of this repository and believe that this is not a duplicate.

Refactor the tag route feature with go.

### PR DESCRIPTION
**Pull request by seriouszyx**
## What is the purpose of the change

Fix: https://github.com/apache/dubbo-admin/issues/1029
Refactor the tag route feature with go

## Verifying this change

Verify by postman:

CreateRule:

```shell
curl -X POST \
  http://127.0.0.1:38080/api/dev/rules/route/tag \
  -H 'cache-control: no-cache' \
  -d '{"configVersion":"v3.0","force":true,"enabled":true,"key":"shop-detail","tags":[{"name":"gray","match":[{"key":"env","value":{"exact":"gray"}}]}],"application":"my-app1"}'
```

UpdateRule:

```shell
curl -X PUT \
  http://127.0.0.1:38080/api/dev/rules/route/tag/my-app1 \
  -H 'cache-control: no-cache' \
  -d '{"configVersion":"v3.1","force":true,"enabled":true,"key":"shop-detail","tags":[{"name":"gray","match":[{"key":"env","value":{"exact":"gray"}}]}],"application":"my-app1"}'
 ```

SearchRoutes:

```bash
curl -X GET \
  'http://127.0.0.1:38080/api/dev/rules/route/tag?application=my-app1' \
  -H 'cache-control: no-cache' \
```

DetailRoute:

```bash
curl -X GET \
  http://127.0.0.1:38080/api/dev/rules/route/tag/my-app1 \
  -H 'cache-control: no-cache' \
```

DeleteRoute:

```bash
curl -X DELETE \
  http://127.0.0.1:38080/api/dev/rules/route/tag/my-app1 \
  -H 'cache-control: no-cache' \
```

EnableRoute:

```bash
curl -X PUT \
  http://127.0.0.1:38080/api/dev/rules/route/tag/enable/my-app1 \
  -H 'cache-control: no-cache' \
```

DisableRoute:

```bash
curl -X PUT \
  http://127.0.0.1:38080/api/dev/rules/route/tag/disable/my-app1 \
  -H 'cache-control: no-cache' \
```

## Issues to be fixed

1. The added unit test relies on the zookeeper environment, so the CI fails.
2. The Update API returns the error "zk: node already exists", which should be caused by the PublishConfig func of the dubbo-go implementation. 

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist.
* [ ] Run `mvn clean compile --batch-mode -DskipTests=false -Dcheckstyle.skip=false -Drat.skip=false -Dmaven.javadoc.skip=true to make sure basic checks pass.

### CODE REVIEW

**Comment by AlbumenJ** [comment]
Please fix conflicts

**Comment by seriouszyx** [comment]
> Please fix conflicts

done

File: pkg/admin/model/tag_route.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+type Tag struct {
+	Name      string `json:"name" binding:"required"`
+	Match     []MatchCondition
+	Addresses []string `json:"addresses"`
+}
+
+type MatchCondition struct {
+	Key   string `json:"key" binding:"required"`
+	Value StringMatch
```
Comment (chickenlj) [inline pkg/admin/model/tag_route.go]: The value field should be 'Value `json:"value" yaml:"value" binding:"required"`'

Below is one rule example
```yaml
configVersion: v3.0
force: true
enabled: true
key: shop-detail
tags:
  - name: gray
    match:
      - key: env
        value:
          exact: gray
```

File: pkg/admin/services/tag_route_service_impl.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	dto.Tags = tagRoute.Tags
+	return dto
+}
+
+func getIdFromDto(baseDto model.TagRouteDto) string {
+	if baseDto.Application != "" {
+		return baseDto.Application
+	}
+	// id format: "${class}:${version}:${group}"
+	return baseDto.Service + constant.Colon + baseDto.ConfigVersion + constant.Colon + baseDto.ServiceGroup
```
Comment (ev1lQuark) [inline pkg/admin/services/tag_route_service_impl.go]: Maybe you should use BuildServiceKey in pkg/admin/util/base_service_metadata.go, I have doubts about the format.

Comment (seriouszyx) [reply pkg/admin/services/tag_route_service_impl.go]: done

Comment (chickenlj) [reply pkg/admin/services/tag_route_service_impl.go]: Instead, below is the right format to use. I think we need to change `BuildServiceKey` to the following format.


```java
// id format: "${class}:${version}:${group}"
return baseDto.Service + constant.Colon + baseDto.ConfigVersion + constant.Colon + baseDto.ServiceGroup
```

**Review by chickenlj — APPROVED** [review]
LGTM.

**Pull request merged**